File: calendar-site/backend/routes/days.js
const express = require('express');
const router = express.Router();
const storage = require('../storage/events');

const MONTH_NAMES = [
  'January', 'February', 'March', 'April', 'May', 'June',
  'July', 'August', 'September', 'October', 'November', 'December'
];

const DAY_NAMES = ['Sunday', 'Monday', 'Tuesday', 'Wednesday', 'Thursday', 'Friday', 'Saturday'];

/**
 * Get events count for a date
 */
async function getEventCountForDate(date) {
  const events = await storage.readEventsForDate(date);
  return events.length;
}

/**
 * Check if a date is today
 */
function isToday(dateStr) {
  const today = new Date();
  const date = new Date(dateStr);
  return date.toDateString() === today.toDateString();
}

/**
 * Check if a date is at month boundary
 */
function isMonthBoundary(dateStr) {
  const date = new Date(dateStr);
  const day = date.getDate();
  const nextDay = new Date(date);
  nextDay.setDate(day + 1);
  return nextDay.getMonth() !== date.getMonth();
}

/**
 * Check if a date is weekend
 */
function isWeekend(dateStr) {
  const date = new Date(dateStr);
  const day = date.getDay();
  return day === 0 || day === 6;
}

/**
 * GET /api/days
 * Get day data for infinite scroll
 * Query params: start=YYYY-MM-DD&count=N
 * Returns: { days: DayMetadata[] }
 */
router.get('/', async (req, res) => {
  try {
    const { start, count } = req.query;
    
    if (!start) {
      return res.status(400).json({ error: 'Missing required parameter: start' });
    }
    
    if (!count) {
      return res.status(400).json({ error: 'Missing required parameter: count' });
    }
    
    const totalCount = parseInt(count, 10);
    if (isNaN(totalCount) || totalCount <= 0) {
      return res.status(400).json({ error: 'count must be a positive number' });
    }
    
    // Validate start date format
    if (!/^\d{4}-\d{2}-\d{2}$/.test(start)) {
      return res.status(400).json({ error: 'Invalid date format. Use YYYY-MM-DD' });
    }
    
    const days = [];
    const startDate = new Date(start);
    
    for (let i = 0; i < totalCount; i++) {
      const currentDate = new Date(startDate);
      currentDate.setDate(startDate.getDate() + i);
      
      const dateStr = currentDate.toISOString().split('T')[0];
      const dayOfMonth = currentDate.getDate();
      const month = currentDate.getMonth() + 1;
      const monthName = MONTH_NAMES[currentDate.getMonth()];
      const dayName = DAY_NAMES[currentDate.getDay()];
      const isWeekendDay = isWeekend(dateStr);
      const isMonthBoundaryDay = isMonthBoundary(dateStr);
      const isTodayFlag = isToday(dateStr);
      const eventCount = await getEventCountForDate(dateStr);
      const hasEvents = eventCount > 0;
      
      days.push({
        date: dateStr,
        dayOfMonth,
        month,
        monthName,
        dayName,
        isWeekend: isWeekendDay,
        isMonthBoundary: isMonthBoundaryDay,
        isToday: isTodayFlag,
        eventCount,
        hasEvents
      });
    }
    
    res.json({ days });
  } catch (error) {
    console.error('Error reading days:', error.message);
    res.status(500).json({ error: 'Failed to retrieve days' });
  }
});

module.exports = router;

File: calendar-site/backend-rust/src/routes/days.rs
// calendar-site/backend-rust/src/routes/days.rs
use axum::extract::{Query, State};
use axum::http::HeaderMap;
use axum::Json;
use serde::Deserialize;
use serde_json::json;
use chrono::{NaiveDate, Datelike, Duration};
use crate::error::AppError;
use crate::models::DayInfo;
use super::super::AppState;
use super::events::require_auth;

const MONTHS: [&str; 12] = [
    "January","February","March","April","May","June",
    "July","August","September","October","November","December"
];
const DAYS: [&str; 7] = ["Sunday","Monday","Tuesday","Wednesday","Thursday","Friday","Saturday"];

#[derive(Deserialize)]
pub struct DaysQuery { pub start: String, pub count: u32 }

pub async fn list(
    State(state): State<AppState>,
    headers: HeaderMap,
    Query(q): Query<DaysQuery>,
) -> Result<Json<serde_json::Value>, AppError> {
    require_auth(&headers, &state).await?;

    if q.count == 0 {
        return Err(AppError::Validation("count must be positive".into()));
    }

    let start = NaiveDate::parse_from_str(&q.start, "%Y-%m-%d")
        .map_err(|_| AppError::Validation("Invalid date format, use YYYY-MM-DD".into()))?;

    let today = chrono::Local::now().date_naive();
    let mut days = Vec::with_capacity(q.count as usize);

    for i in 0..q.count {
        let d = start + Duration::days(i as i64);
        let date_str = d.format("%Y-%m-%d").to_string();
        let dow = d.weekday().num_days_from_sunday() as usize; // 0=Sun
        let is_weekend = dow == 0 || dow == 6;

        // month boundary: last day of month (next day is a different month)
        let next = d + Duration::days(1);
        let is_month_boundary = next.month() != d.month();

        let event_count = state.events.count_for_date(&date_str).await;

        days.push(DayInfo {
            date: date_str,
            day_of_month: d.day(),
            month: d.month(),
            month_name: MONTHS[(d.month0()) as usize].to_string(),
            day_name: DAYS[dow].to_string(),
            is_weekend,
            is_month_boundary,
            is_today: d == today,
            event_count,
            has_events: event_count > 0,
        });
    }

    Ok(Json(json!({ "days": days })))
}
